File: sdk/php/src/Exception/CaptchaException.php
<?php

namespace HJTPX\Captcha\Exception;

use Exception;

class CaptchaException extends Exception
{
    protected $errorCode;
    protected $context;

    public function __construct(
        string $message = '',
        int $code = 0,
        ?\Throwable $previous = null,
        string $errorCode = null,
        array $context = []
    ) {
        parent::__construct($message, $code, $previous);
        $this->errorCode = $errorCode;
        $this->context = $context;
    }

    public function getErrorCode(): ?string
    {
        return $this->errorCode;
    }

    public function getContext(): array
    {
        return $this->context;
    }

    public function toArray(): array
    {
        return [
            'message' => $this->getMessage(),
            'code' => $this->getCode(),
            'error_code' => $this->errorCode,
            'context' => $this->context,
            'file' => $this->getFile(),
            'line' => $this->getLine(),
        ];
    }
}

File: sdk/java/src/main/java/com/hjtpx/captcha/exception/CaptchaException.java
package com.hjtpx.captcha.exception;

public class CaptchaException extends RuntimeException {
    private final String code;
    private final boolean retryable;

    public CaptchaException(String message) {
        super(message);
        this.code = null;
        this.retryable = false;
    }

    public CaptchaException(String message, String code) {
        super(message);
        this.code = code;
        this.retryable = false;
    }

    public CaptchaException(String message, String code, boolean retryable) {
        super(message);
        this.code = code;
        this.retryable = retryable;
    }

    public CaptchaException(String message, Throwable cause) {
        super(message, cause);
        this.code = null;
        this.retryable = false;
    }

    public CaptchaException(String message, String code, Throwable cause) {
        super(message, cause);
        this.code = code;
        this.retryable = false;
    }

    public CaptchaException(String message, String code, boolean retryable, Throwable cause) {
        super(message, cause);
        this.code = code;
        this.retryable = retryable;
    }

    public String getCode() {
        return code;
    }

    public boolean isRetryable() {
        return retryable;
    }
}
